cli/src/commands/evolog.rs

```diff
@@ -151,14 +158,53 @@ pub(crate) fn cmd_evolog(
     if !args.no_graph {
         let mut raw_output = formatter.raw()?;
         let mut graph = get_graphlog(graph_style, raw_output.as_mut());
-        for commit in commits {
-            let edges = commit
-                .predecessor_ids()
-                .iter()
-                .map(|id| Edge::Direct(id.clone()))
-                .collect_vec();
+
+        let graph_nodes: Vec<Result<GraphNode<Commit>, ()>> = commits
+            .iter()
+            .map(|c| {
+                let edges = c
+                    .predecessors()
+                    // TODO: Figure out what to do with this unwrap, if this approach works.
+                    .map(|p| GraphEdge::direct(p.unwrap().clone()))
+                    .collect_vec();
+                Ok((c.clone(), edges))
+            })
+            .collect_vec();
+
+        // TODO: This exists to strip out the Result types because ReverseGraphIterator
+        // assumes that the error type of its input iterator is RevsetEvaluationError
+        // when in this case, the input iterator is infallible (E is `()`), or at the
+        // very least a different error type. This is clunky but monkeying around in
+        // lib/src/graph proved to be a bit much for me.
```

Comment:
I don't follow the underlying problem, but `ReverseGraphIterator` isn't really a (lazy) iterator. Maybe we can extract a function that consumes inputs and turns them into `Vec<GraphNode>>`.

Reply:
To try and explain where I got stuck more clearly: I made the items in `graph_nodes` be `Result<GraphNode<Commit>, ()>` so that I could pass it to `ReverseGraphIterator::new`. But, because of the `Iterator` implementation for `ReverseGraphIterator`, its `Item` type is `Result<GraphNode<Commit>, RevsetEvaluationError>`.

Since `()` (or whatever other error it might be) doesn't match `RevsetEvaluationError`, I can't have a single value that could be either the normal or reversed iterator (like how I did when `--no-graph` is passed).

I tried typing `graph_nodes` as `Vec<Result<GraphNode<Commit>, RevsetEvaluationError>`, but then I ran into issues where `RevsetEvaluationError` doesn't implement `Clone`.

Alternately, I tried re-tooling the `Iterator` implementation so that `RevsetEvaluationError` isn't hard-coded, but I couldn't figure out how to do it. At least without making more changes than I was comfortable with as a 1st time contributor.

---

Maybe I'll take a fresh look at this code tomorrow / in the next few days and try to clean it up again. If I get stuck, I can come back with more specific blockers.

Reply:
> extract a function that consumes inputs and turns them into `Vec<GraphNode>>`
https://github.com/jj-vcs/jj/pull/5172

Reply:
This makes sense to me! I'm not sure what my availability will be like going into Christmas but I'll report back with updated changes when I have time.